storagemarket/provider_test.go

```diff
@@ -0,0 +1,552 @@
+package storagemarket
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"math/rand"
+	"net/http/httptest"
+	"os"
+	"path/filepath"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/filecoin-project/boost/storagemarket/smtestutil"
+
+	"github.com/libp2p/go-libp2p-core/event"
+
+	carv2 "github.com/ipld/go-car/v2"
+
+	"github.com/ipfs/go-cid"
+
+	"github.com/filecoin-project/boost/storagemarket/types/dealcheckpoints"
+
+	"github.com/filecoin-project/go-state-types/big"
+
+	"github.com/filecoin-project/lotus/api"
+
+	types2 "github.com/filecoin-project/boost/transport/types"
+
+	acrypto "github.com/filecoin-project/go-state-types/crypto"
+
+	"github.com/filecoin-project/boost/testutil"
+	"github.com/google/uuid"
+
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/specs-actors/actors/builtin/market"
+
+	"github.com/filecoin-project/boost/db"
+	"github.com/filecoin-project/boost/fundmanager"
+	"github.com/filecoin-project/boost/storagemanager"
+	"github.com/filecoin-project/boost/storagemarket/types"
+	"github.com/filecoin-project/go-address"
+	lotusmocks "github.com/filecoin-project/lotus/api/mocks"
+
+	"testing"
+
+	"github.com/filecoin-project/lotus/node/repo"
+	"github.com/golang/mock/gomock"
+	"github.com/libp2p/go-libp2p-core/peer"
+	mocknet "github.com/libp2p/go-libp2p/p2p/net/mock"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSimpleDealHappy(t *testing.T) {
+	ctx := context.Background()
+	testMode = true
+	defer func() {
+		testMode = false
+	}()
```

Comment:
I left a comment suggesting how we can avoid using a testMode variable.
If we can't avoid testMode, I'd suggest adding it as a field on Provider, so it gets set up automatically in the harness instead of needing to add this code to each test.

Reply:
It's hard to avoid that variable for now because navigating the call hierarchy of the validator will need some grunt work (see what we do in `go-fil-markets` right now, we basically mock out the entire call hierarchy). For now, I am moving this variable as a field on the Provider as you have suggested.

Reply:
As a part of writing the tests, I have enlisted a testcase for all the validation logic as well.